wandb/sdk/lib/sock_client.py

```diff
@@ -51,7 +52,14 @@ def _send_message(self, msg: Any) -> None:
         data = msg.SerializeToString()
         assert len(data) == raw_size, "invalid serialization"
         header = struct.pack("<BI", ord("W"), raw_size)
-        self._sock.sendall(header + data)
+        total_sent = 0
+        message = header + data
+        while total_sent < len(message):
+            with self._lock:
+                sent = self._sock.send(message[total_sent:])
+            if sent == 0:
+                raise RuntimeError("socket connection broken")
```

Comment:
I wonder if it's worth handling retry / catching other socket errors that could be transient..,

Reply:
oh yeah... I added support to handle retry in case of timeout. Are there any other failures you are worried about?

Reply:
I dont think there is much more, retry is handled internally as of python3.? +

Maybe 35 ERESOURCE.. the out of resource one.. you could probably test this by setting your socket buffer very low.

If you add more, it should be well reasoned or tested

Reply:
I think it is valid to except in this case since we got the system in a bad state?